lib/Sema/TypeCheckConcurrency.cpp

```diff
@@ -338,6 +338,69 @@ class IsolationRestriction {
 
 }
 
+namespace {
+  /// Describes the important parts of a partial apply thunk.
+  struct PartialApplyThunkInfo {
+    Expr *base;
+    Expr *fn;
+    bool isEscaping;
+  };
+}
+
+/// Try to decompose a call that might be an invocation of a partial apply
+/// thunk.
+static Optional<PartialApplyThunkInfo> decomposePartialApplyThunk(
+    ApplyExpr *apply, Expr *parent) {
+  // Check for a call to the outer closure in the thunk.
+  auto outerAutoclosure = dyn_cast<AutoClosureExpr>(apply->getFn());
+  if (!outerAutoclosure ||
```

Comment:
I think you can use `getUnwrappedCurryThunkExpr()` on `outerAutoclosure` to dig out the `DeclRefExpr` of the function.

Reply:
I didn't know this function existed! I'll figure out how to use it, thank you.

Reply:
Heh, good news. Doing do improved test coverage for `getUnwrappedCurryThunkExpr()`, so now both my new client and all existing clients handle more curry think ASTs properly.